### PR TITLE
✨Return all NodeAddresses in OpenStackMachine.Status.Addresses

### LINKED ISSUE #926: Instance.IP is set randomly when a server has multiple networks
/kind bug

This is from code inspection while working on MAPO: https://github.com/shiftstack/machine-api-provider-openstack
There is no reproducer, I could have misread the code, and/or this may not be relevant to CAPO.

The problem is where we're iterating over a server's Addresses in here: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/master/pkg/cloud/services/compute/instance.go#L916-L930
```go
		for _, network := range networkList {
			var netInterface networkInterface
			b, _ := json.Marshal(network)
			err = json.Unmarshal(b, &netInterface)
			if err != nil {
				return nil, fmt.Errorf("extract IP from instance err: %v", err)
			}
			if netInterface.Version == 4.0 {
				if netInterface.Type == "floating" {
					addrMap["floating"] = netInterface.Address
				} else {
					addrMap["internal"] = netInterface.Address
				}
			}
		}
```

If the server has multiple networks we will return the last one returned by the API which is not a floating IP as the machine's control plane IP.

The problem here is that as this interface is currently used it doesn't have enough context to determine which network of multiple networks is the control plane network. When we call GetInstance() or InstanceExists() we're passing in an OpenStack identifier. However, to determine which is the control plane IP we additionally need to know which is the control plane network. AFAICT the canonical reference for this would be OpenStackMachine.Spec.Subnet.

I'm currently thinking of a refactor. I think serverToInstance() needs this additional context. I also note that GetIPFromInstance is called by an E2E test, and duplicates a bunch of code. I wonder if it might be cleaner to have it use an exported serverToInstance instead. I haven't got a firm plan in mind, yet.

### LINKED ISSUE #1001: Report all machine addresses
/kind feature

**Describe the solution you'd like**
CAPO should report *all* addresses of a machine in OpenstackMachine.Status.Addresses, not just a single internal IP and a single external IP.

This would make the behaviour of CAPO consistent with the only 2 other providers I checked, AWS and GCP, both of which report all addresses.

*AWS:*
reconcileNormal:
https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/151fb40cbaa2d87d846f4a0f620c2dbf8b440283/controllers/awsmachine_controller.go#L600

SDKToInstance:
https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/151fb40cbaa2d87d846f4a0f620c2dbf8b440283/pkg/cloud/services/ec2/instances.go#L780

getInstanceAddresses:
https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/151fb40cbaa2d87d846f4a0f620c2dbf8b440283/pkg/cloud/services/ec2/instances.go#L791-L818

*GCP*
Reconcile:
https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/582ef8e5548ee0357689dd9987bf70ac3839695d/cloud/services/compute/instances/reconcile.go#L45-L61

**Anything else you would like to add:**
Another motivation for doing this is removing a use of the problematic instanceStatus.NetworkStatus().IP() and FloatingIP() methods.

### PR DESCRIPTION
**Pull request by mdbooth**
This commit makes 2 closely related changes:
* Add NetworkStatus.Addresses(), and return this in OpenStackMachine
* Update IP() and FloatingIP() to return individual addresses by
  network name

There are also a couple of incidental changes:
* Don't bother with DisassociateFloatingIP before DeleteFloatingIP for
  bastion host
* Delete floating IPs before server during machine delete: avoids
  orphaned floating IPs if floating IP delete fails.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1001, fixes #926

### CODE REVIEW

**Comment by mdbooth** [comment]
Multiple addresses: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-openstack/1004/pull-cluster-api-provider-openstack-e2e-test/1440688542236807168/artifacts/clusters/bootstrap/resources/e2e-0vpq66/Machine/cluster-e2e-0vpq66-control-plane-28crg.yaml

File: controllers/openstackcluster_controller.go [hunk 1]
```diff
@@ -206,16 +207,14 @@ func deleteBastion(log logr.Logger, osProviderClient *gophercloud.ProviderClient
...
-		floatingIP := instanceNS.FloatingIP()
-		if floatingIP != "" {
-			if err = networkingService.DisassociateFloatingIP(openStackCluster, floatingIP); err != nil {
-				handleUpdateOSCError(openStackCluster, errors.Errorf("failed to disassociate floating IP: %v", err))
-				return errors.Errorf("failed to disassociate floating IP: %v", err)
-			}
-			if err = networkingService.DeleteFloatingIP(openStackCluster, floatingIP); err != nil {
-				handleUpdateOSCError(openStackCluster, errors.Errorf("failed to delete floating IP: %v", err))
-				return errors.Errorf("failed to delete floating IP: %v", err)
+		for _, address := range addresses {
```
Comment (jichenjc) [inline controllers/openstackcluster_controller.go:212]: just curious, what's the reason behind the removal of disasociate floating ip for bastion only?

File: controllers/openstackmachine_controller.go [hunk 2]
```diff
@@ -224,28 +224,27 @@ func (r *OpenStackMachineReconciler) reconcileDelete(ctx context.Context, logger
 
 	if instanceStatus == nil {
 		logger.Info("Skipped deleting machine that is already deleted")
```
Comment (jichenjc) [inline controllers/openstackmachine_controller.go:226]: maybe add instance ID here?

File: pkg/cloud/services/compute/instance_types.go [hunk 3]
```diff
@@ -132,51 +135,100 @@ func (is *InstanceStatus) InstanceIdentifier() *InstanceIdentifier {
...
-			if vif.Version == 4.0 && vif.Type != "floating" {
-				return vif.Address
+			var addressType corev1.NodeAddressType
+			switch address.Type {
+			case "floating":
+				addressType = corev1.NodeExternalIP
+			case "fixed":
+				addressType = corev1.NodeInternalIP
+			default:
+				// Ignore unknown address types
```
Comment (jichenjc) [inline pkg/cloud/services/compute/instance_types.go:174]: maybe add some log as reference? though should not reach here..

File: pkg/cloud/services/compute/instance_types.go [hunk 4]
```diff
@@ -132,51 +135,100 @@ func (is *InstanceStatus) InstanceIdentifier() *InstanceIdentifier {
...
 }
+
+// IP returns the first listed ip of an instance for the given network name.
+func (ns *InstanceNetworkStatus) IP(networkName string) string {
+	return ns.firstAddressByNetworkAndType(networkName, corev1.NodeInternalIP)
+}
+
+// FloatingIP returns the first listed floating ip of an instance for the given
+// network name.
+func (ns *InstanceNetworkStatus) FloatingIP(networkName string) string {
```
Comment (jichenjc) [inline pkg/cloud/services/compute/instance_types.go:232]: I guess I lost some background
usually ,multiple network is reasonable, so we have multiple NICs from various network
but multiple NIC from one network is not so familiar to me, anyway, seems no harm if we 
return the first ip

**Comment by jichenjc** [comment]
/lgtm

overall looks great, just very some minor questions.. thanks~

File: controllers/openstackcluster_controller.go (hunk 1 above)
Comment (mdbooth) [reply controllers/openstackcluster_controller.go:212]: Consistency. We've never done it for controllers: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/cf5e5cb826db47cf79ccd3d8bdd2a779f3c584c0/controllers/openstackmachine_controller.go#L246-L251

Also it's redundant, because the floating IP will be disassociated automatically on deletion.

Comment (jichenjc) [reply controllers/openstackcluster_controller.go:212]: ok, thanks for the confirmation
I don't know the disassociate also happen on deletion.. thanks:)

File: pkg/cloud/services/compute/instance_types.go (hunk 3 above)
Comment (mdbooth) [reply pkg/cloud/services/compute/instance_types.go:174]: I thought about this, and I'm not sure.

The most likely reason that we would get an unknown address type here is that OpenStack introduced a new address type which we don't use, yet. This would be a new feature, and logging something here because of a new feature we don't support yet would be very noisy. I guess we could log this and IPv6 addresses at debug level just in case?

Interestingly, I was speaking to a Nova developer about `addresses` yesterday, and his impression is that `addresses` is more likely to go away entirely rather than gain new features because it's viewed as a proxy API. The replacement would be to directly query neutron for ports associated with the device id. The only reason I didn't make that change here is that as long as Nova is caching NetworkInfo, getting it from Nova requires 1 less API call.

TL;DR I might add this at debug.

File: pkg/cloud/services/compute/instance_types.go (hunk 4 above)
Comment (mdbooth) [reply pkg/cloud/services/compute/instance_types.go:232]: We're actually only using this method in 1 place: APIInstance() above. And the only thing that uses APIInstance() is the bastion host. I guess we're not likely to put a second floating IP on the bastion host, so this is really just something to make the behaviour defined if we ever did.

File: pkg/cloud/services/compute/instance_types.go (hunk 3 above)
Comment (mdbooth) [reply pkg/cloud/services/compute/instance_types.go]: I've added this.

File: controllers/openstackmachine_controller.go (hunk 2 above)
Comment (mdbooth) [reply controllers/openstackmachine_controller.go:226]: I added this, but then removed it again because after looking at how we create the log object I realised it's redundant. We don't have instance ID here because the machine has been deleted. I had added machine name, but that's already included by the logger: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/cf5e5cb826db47cf79ccd3d8bdd2a779f3c584c0/controllers/openstackmachine_controller.go#L88

File: test/e2e/suites/e2e/e2e_test.go [hunk 5]
```diff
@@ -192,6 +198,59 @@ var _ = Describe("e2e tests", func() {
...
+				Namespace:         namespace.Name,
+				MachineDeployment: *md[0],
+			})
+			controlPlaneMachines := framework.GetControlPlaneMachinesByCluster(ctx, framework.GetControlPlaneMachinesByClusterInput{
+				Lister:      e2eCtx.Environment.BootstrapClusterProxy.GetClient(),
+				ClusterName: clusterName,
+				Namespace:   namespace.Name,
+			})
+			Expect(len(workerMachines)).To(Equal(1))
+			Expect(len(controlPlaneMachines)).To(Equal(3))
```
Comment (macaptain) [inline test/e2e/suites/e2e/e2e_test.go:250]: I see that this creates a four node cluster with the machines having ports on two different networks. Did you also want to test your functions? I think you should be able to see e.g. `controlPlaneMachines[0].Status.Addresses`, if the controller is working, this should show multiple addresses, since they get assigned in reconciliation: 
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/fcbf20746b7dfe3b7776c1d7420c341e1225d9cb/controllers/openstackmachine_controller.go#L334

**Review by macaptain — APPROVED** [review]
Looks great! Unit tests and e2e tests, we are spoiled.

File: pkg/cloud/services/compute/instance_types.go [hunk 6]
```diff
@@ -132,51 +137,101 @@ func (is *InstanceStatus) InstanceIdentifier() *InstanceIdentifier {
...
+func (ns *InstanceNetworkStatus) Addresses() []corev1.NodeAddress {
+	// We want the returned order of addresses to be deterministic to make
+	// it easy to detect changes and avoid unnecessary updates. Iteration
+	// over maps is non-deterministic, so we explicitly iterate over the
+	// address map in lexical order of network names. This order is
+	// arbitrary.
+	networks := make([]string, 0, len(ns.addresses))
+	for network := range ns.addresses {
+		networks = append(networks, network)
+	}
```
[the remark this answers is not in the document]
Comment (mdbooth) [reply pkg/cloud/services/compute/instance_types.go:204]: I read your comment, looked at my code, and agreed with you.

I was trying to work out why I wrote it this way when I realised the code is actually correct, because the range iterates over only the keys of the map, not the elements.

I think that's a fairly clear indication this code needs another comment 😂 I'll add one.

File: test/e2e/suites/e2e/e2e_test.go (hunk 5 above)
Comment (mdbooth) [reply test/e2e/suites/e2e/e2e_test.go:250]: This was a lazy late addition, tbh. I'd previously written it as a standalone create/delete test which I expected to fail. However, it turns out that it eventually comes up because although IP selection was non-deterministic and failed when it picked the wrong one, the reconciler would just retry and eventually it picked the right one 🤷‍♂️

You are right: this test would be massively improved by testing the resulting addresses.

File: pkg/cloud/services/compute/instance_types.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (mdbooth) [reply pkg/cloud/services/compute/instance_types.go:205]: Added a comment.

File: test/e2e/suites/e2e/e2e_test.go (hunk 5 above)
Comment (mdbooth) [reply test/e2e/suites/e2e/e2e_test.go]: I've expanded the e2e test to explicitly check ports and addresses.

Incidentally, @EmilienM you may want to do something like this in #1008? I'm iterating over openstack ports per machine, which I think it what we were trying to do in that test.

**Review by macaptain — APPROVED** [review]

**Review by chrischdi — COMMENTED** [review]
/lgtm
